api/core/helper/ssrf_proxy.py

```diff
@@ -15,34 +17,56 @@
 } if SSRF_PROXY_HTTP_URL and SSRF_PROXY_HTTPS_URL else None
 
 
-def make_request(method, url, **kwargs):
-    if SSRF_PROXY_ALL_URL:
-        return httpx.request(method=method, url=url, proxy=SSRF_PROXY_ALL_URL, **kwargs)
-    elif proxies:
-        return httpx.request(method=method, url=url, proxies=proxies, **kwargs)
-    else:
-        return httpx.request(method=method, url=url, **kwargs)
+DEFAULT_MAX_RETRIES = 0
+BACKOFF_FACTOR = 0.5
+STATUS_FORCELIST = [429, 500, 502, 503, 504]
 
 
-def get(url, **kwargs):
-    return make_request('GET', url, **kwargs)
+def make_request(method, url, max_retries=DEFAULT_MAX_RETRIES, **kwargs):
+    retries = 0
+    while retries <= max_retries:
+        try:
+            if SSRF_PROXY_ALL_URL:
+                response = httpx.request(method=method, url=url, proxy=SSRF_PROXY_ALL_URL, **kwargs)
+            elif proxies:
+                response = httpx.request(method=method, url=url, proxies=proxies, **kwargs)
+            else:
+                response = httpx.request(method=method, url=url, **kwargs)
 
+            if response.status_code not in STATUS_FORCELIST:
+                return response
+            else:
+                logging.warning(f"Received status code {response.status_code} for URL {url} which is in the force list")
 
-def post(url, **kwargs):
-    return make_request('POST', url, **kwargs)
+        except httpx.RequestError as e:
+            logging.warning(f"Request to URL {url} failed on attempt {retries + 1}: {e}")
 
+        retries += 1
+        if retries <= max_retries:
+            time.sleep(BACKOFF_FACTOR * (2 ** (retries - 1)))
 
-def put(url, **kwargs):
-    return make_request('PUT', url, **kwargs)
+    return None
```

Comment:
An exception need to be raised here I think, otherwise it's a little subtle, as the caller, I need an exception to know that it has reached max_retries.